src/NuGet.Core/NuGet.Packaging/Rules/ReferencesInNuspecMatchRefAssetsRule.cs

```diff
@@ -0,0 +1,137 @@
+
+        internal IEnumerable<MissingReference> Compare(IDictionary<string, string[]> nuspecReferences, IEnumerable<string> refFiles)
+        {
+            var filesByTFM = refFiles.Where(t => t.Count(m => m == '/') > 1).GroupBy(t => NuGetFramework.ParseFolder(t.Split('/')[1]).GetShortFolderName(), t => GetFile(t));
+            var missingItems = new List<MissingReference>();
+
+            foreach (var files in filesByTFM)
+            {
+                if(files.Key == "unsupported")
+                {
+                    continue;
+                }
+
+                var missingFiles = Array.Empty<string>();
+                var missingReferences = Array.Empty<string>();    
+                if (nuspecReferences.TryGetValue(files.Key, out var currentReferences))
+                {
+                    missingReferences = files.Where(m => !currentReferences.Contains(m)).ToArray();
+                    missingFiles = currentReferences.Where(t => !files.Contains(t)).ToArray();
+                }
+                else
+                {
+                    missingReferences = files.ToArray();
+                }
+
+                if (missingFiles.Length != 0)
+                {
+                    missingItems.Add(new MissingReference("ref", files.Key, missingFiles));
+                }
+
+                if (missingReferences.Length != 0)
+                {
+                    missingItems.Add(new MissingReference("nuspec", files.Key, missingReferences));
+                }
+            }
+            return missingItems;
+        }
+
+        internal IEnumerable<PackagingLogMessage> GenerateWarnings(IEnumerable<MissingReference> missingItems)
+        {
+            var issues = new List<PackagingLogMessage>();
+            if (missingItems.Count() != 0)
+            {
+                var message = new StringBuilder();
+                message.AppendLine(MessageFormat);
+                var referencesMissing = missingItems.Where(t => t.MissingFrom == "nuspec");
+                var refFilesMissing = missingItems.Where(t => t.MissingFrom == "ref");
+                foreach (var file in refFilesMissing)
+                {
+                    foreach(var item in file.MissingItems)
+                    {
+                        message.AppendLine(string.Format(_addToRefFormat, item, file.TFM));
+                    }
+                }
+
+                foreach (var reference in referencesMissing)
+                {
+                    foreach (var item in reference.MissingItems)
+                    {
+                        message.AppendLine(string.Format(_addToNuspecFormat, item, reference.TFM));
+                    }
+                }
+                issues.Add(PackagingLogMessage.CreateWarning(message.ToString(), NuGetLogCode.NU5131));
+            }
+            return issues;
+        }
+
+        private string GetFile(string filePath)
+        {
+            return filePath.Split('/')[filePath.Count(p => p == '/')];
```

Comment:
```csharp
var segments = filePath.Split('/');
return segments[segments.Length - 1];
```

would have higher performance than parsing `filePath` twice. But `Path.GetFileName(string)` already exists, so you don't need this helper method.